File: benchmark.cpp
#include <chrono>
#include <cmath>
#include <cstdio>
#include <vector>

typedef std::chrono::high_resolution_clock Clock;

double bench(const char* name, long long iters, auto fn) {
    for (int w = 0; w < 3; w++) fn();
    auto t0 = Clock::now();
    fn();
    auto t1 = Clock::now();
    double ms = std::chrono::duration<double, std::milli>(t1 - t0).count();
    double ns = ms * 1000000.0 / iters;
    printf("  %-30s  %8.1f ns/op   %8.3f ms\n", name, ns, ms);
    return ns;
}

int main() {
    printf("C++ Benchmark (-O0, naive)\n\n");

    const long long N1 = 10000000;
    bench("int counter loop       10M", N1, [N1]() {
        long long s = 0;
        for (long long i = 0; i < N1; i++) { s += i; }
        return s;
    });

    const long long N2 = 5000000;
    bench("float mul/add           5M", N2, [N2]() {
        double s = 1.0;
        for (long long i = 1; i < N2; i++) {
            s = s * 1.0000001 + 0.0000001;
        }
        return s;
    });

    const long long N3 = 2000000;
    bench("math.sqrt               2M", N3, [N3]() {
        double s = 0.0;
        for (long long i = 1; i < N3; i++) { s += std::sqrt(i); }
        return s;
    });

    const long long N4 = 1000000;
    bench("sin + cos               1M", N4, [N4]() {
        double s = 0.0;
        for (long long i = 0; i < N4; i++) { s += std::sin(i) + std::cos(i); }
        return s;
    });

    const long long N8 = 100000;
    bench("array push+sum        100K", N8, [N8]() {
        std::vector<long long> arr;
        for (long long i = 0; i < N8; i++) { arr.push_back(i * i); }
        long long s = 0;
        for (long long j = 0; j < N8; j++) { s += arr[j]; }
        return s;
    });

    const long long N9 = 500000;
    auto add2 = [](long long a, long long b) { return a + b; };
    bench("fn call overhead      500K", N9, [N9, &add2]() {
        long long s = 0;
        for (long long i = 0; i < N9; i++) { s = add2(s, i); }
        return s;
    });

    const long long N10 = 1000;
    bench("nested loops           1M", N10 * N10, [N10]() {
        long long s = 0;
        for (long long i = 0; i < N10; i++) {
            for (long long j = 0; j < N10; j++) { s += i * j; }
        }
        return s;
    });

    printf("\n");
    return 0;
}

File: benchmark.py
import time, math

def bench(name, iters, fn):
    for _ in range(3): fn()
    t0 = time.perf_counter()
    fn()
    t1 = time.perf_counter()
    ms = (t1 - t0) * 1000
    ns = ms * 1000000 / iters
    print(f"  {name:30s}  {ns:8.1f} ns/op   {ms:8.3f} ms")

print("Python 3 Benchmark")
print("")

N1 = 10_000_000
def f1():
    s = 0
    for i in range(N1):
        s += i
    return s
bench("int counter loop       10M", N1, f1)

N2 = 5_000_000
def f2():
    s = 1.0
    for i in range(1, N2):
        s = s * 1.0000001 + 0.0000001
    return s
bench("float mul/add           5M", N2, f2)

N3 = 2_000_000
def f3():
    s = 0.0
    for i in range(1, N3):
        s += math.sqrt(i)
    return s
bench("math.sqrt               2M", N3, f3)

N4 = 1_000_000
def f4():
    s = 0.0
    for i in range(N4):
        s += math.sin(i) + math.cos(i)
    return s
bench("sin + cos               1M", N4, f4)

N8 = 100_000
def f8():
    arr = []
    for i in range(N8):
        arr.append(i * i)
    s = 0
    for j in range(N8):
        s += arr[j]
    return s
bench("array push+sum        100K", N8, f8)

N9 = 500_000
def add2(a, b):
    return a + b
def f9():
    s = 0
    for i in range(N9):
        s = add2(s, i)
    return s
bench("fn call overhead      500K", N9, f9)

N10 = 1000
def f10():
    s = 0
    for i in range(N10):
        for j in range(N10):
            s += i * j
    return s
bench("nested loops           1M", N10 * N10, f10)

print("")
